### DISCUSSION TITLE
Provide facet distribution of tasks

### DISCUSSION THREAD

**Discussion by bidoubiwa** [meilisearch/product#654] [Feedback & Feature Proposal]
## Suggestion

When requesting tasks on the `GET /tasks` route, we are lacking knowledge on the number of tasks. 
In #635 , we talk about returning a simple `totalTask` field, which provides the total number of tasks left after the filter are applied.

In this discussion, I would like to suggest that we provide a `facetDistribution` or `stats` object that provides more information on the total tasks.

This `stats` object may either be on its own route `GET /tasks/stats` or added in the response of `GET /tasks` route.

The stats object would require at least the `filter` payload as it should reflect the numbers based on the filters.

In the example below, I'll use `stats` but at the end of the day, this is a facet distribution.

For example, on a simple call without any filter:

```json
{
  "stats": {
       "byIndexes": {
            "movies": 10,
            "movies_review": 5,
       },
      "byStatuses": {
            "enqueued": 1,
            "processing": 2,
            "failed": 2,
            "succeeded": 10,
            "canceled": 0
       },
       "byTypes": {
            "documentAdditionOrUpdate": 10,
            "taskDeletion": 3,
            "taskCancelation": 1,
            "dumpCreation": 1,
            // etc...
       },
       "canceled": 0
}
```

On another call with the following payload:
```json
{
   "filter": {
        "indexUids": ["movies"]
   }
}

```

The stats would be transformed like this:

```json
{
  "stats": {
       "byIndexes": {
            "movies": 10,
            "movies_review": 0,
       },
      "byStatuses": {
            "enqueued": 1,
            "processing": 2,
            "failed": 2,
            "succeeded": 5, // removed the 5 movies_review that were successful
            "canceled": 0
       },
       "byTypes": {
            "documentAdditionOrUpdate": 5, // removed the 5 movies_review that were document additions
            "taskDeletion": 3,
            "taskCancelation": 1,
            "dumpCreation": 1,
            // etc...
       },
       "canceled": 0
}
```

## Use case

This provides the possibility to showcase the facetDistribution of tasks.

In the following image, we want the distribution of tasks in the statuses based on two other filters: the movies tasks in the last 24h (only look at the distribution not the list of tasks behind). 

Example:

<img width="1052" alt="Screenshot 2023-05-17 at 17 09 18" src="https://github.com/meilisearch/product/assets/33010418/10aefd5a-4f6e-4ffe-aa2b-ae4ec7d9322e">

See for example how two document deletion tasks happened in the last 24 hours on the `movies` index.

## Alternative

In the image above, you can observe that if the filter includes a specific indexUid, the other indexUids still have a number.
If you look at my JSON examples, in the same scenario the other index uid `movie_rating` is set at 0.

In the first case, it uses a facet distribution practice that we can often observe in e-commerce, where a facet does not apply its own filters.

Example:
<img width="1073" alt="Screenshot 2023-05-17 at 16 39 32" src="https://github.com/meilisearch/product/assets/33010418/fdffa6e5-d592-4b72-b436-f8c41bf7abc2">

In the second case, it uses the default behavior of the `AND` operator.
Example:

<img width="1077" alt="Screenshot 2023-05-17 at 16 39 52" src="https://github.com/meilisearch/product/assets/33010418/9b949568-e9a3-4761-a09f-6d1116112cce">

While the simplest implementation would be the second case, the first case might be considered for implementation at Meilisearch side.

What do you think?

**Comment by Kerollmops** [comment]
I would prefer that we expose the task distribution stats from another route. This way, we can keep having those stats, specify the filters on it too, and it doesn't complicate the `/tasks` route too much. However, I am not against adding a `total_tasks` field to the returned object. The Cloud or a few user products will use the `/tasks/stats` route. We shouldn't add that info to the `/tasks` route. It can grow a lot.

**Reply by bidoubiwa** [reply]
Would there be a possibility to provide `filters` to the stats route? Since they need it for their use case
